[Mohamad Shafik] iP

AD-NAP opened the pull request:
# McBot
> Monitor your task like never before 

McBot can handle three different types of tasks: 
- ToDo
- Deadline
- Event

Why should you use it?
1. It's **FREE**
2. Easy to use
3. Efficient for ~~everyone~~ fast typist ⏩ 
#
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
Features: 
- [x] Different types of task
- [ ] Custom font
- [ ] Search tasks

[Click me](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Commit:
Add Gradle support

Commit:
Level 1. Implemented a skeletal version of McBot, previously known as Duke.

Commit:
Level 2. Added the ability to store text entered by the user and list it back.

Commit:
Added a Task class and option to mark tasks as done or change it back to undone.

Commit:
Add support for tracking three types of tasks:
1. Todo - task without dates
2. Deadline - task with a due date/time
3. Event - task with a start and end time

Commit:
Modified input.txt to input each type of tasks, mark and unmark operations. Modified EXPECTED.TXT to have the expected output.

Commit:
Level-5. Added error handling in McBot by using McBotException class. Updated input and expected text files to test all possible error

Commit:
Level-6. Added support for deleting tasks from the list. Modified input and expected text for testing of deletion

Commit:
Add invalidCommandException.java and minor refactoring of the codes.

Commit:
Add save and load feature

Users now able to save their tasks list in a text file whenever the task list changes. The same file is loaded when McBot is started

Refactored redundant methods in ToDo, Deadline and Event classes. Utilise toString method to print out the task details.

Add toDataString in Task type classes for consistent formatting when saving into data text file.

Commit:
Add date and time field for Deadline and Event classes.

The date field must be filled while time is optional.

Commit:
Revert "Add date and time field for Deadline and Event classes."

This reverts commit 2477b6b7014efa48030554eedf147c87a18dbff5.

Commit:
Add date and time field for Deadline and Event classes.

The date field must be filled while time is optional.

Commit:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save and load feature

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add date and time field for Deadline and Event classes.

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/McBot.java

Resolve conflict where Level-8 branch did not have the updated LocalDate and LocalTime fields from Level-7 branch

Commit:
Fix bug in saving and loading time aspect of Deadline and Event tasks

Time field was not saved into the text file, neither could it load properly.

Commit:
Add more OOP classes

By extracting Ui, Storage, Parser and TaskList classes from McBot class, more OOP can be used in this iP. More OOP != good OOP

Edit input and expected text files so it can test the new commands and errors thrown.

Commit:
Add mcbot package which is further divided into task and exception packages.

Refactor invalidCommandException.java to InvalidCommandException.java

Edit runtest.sh so it will work with the packages

Commit:
Add test folder and files to test Task and ToDo classes through junit testing

Commit:
Add Java documentation to the each type of Tasks, Exceptions and Storage files.

Commit:
Tweak the code to comply with a coding standard

Commit:
Add find feature to McBot

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add Java documentation to the each type of Tasks, Exceptions and Storage files.

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with a coding standard

# Conflicts:
#	src/main/java/mcbot/task/ToDo.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find feature to McBot

# Conflicts:
#	src/main/java/mcbot/McBot.java
#	src/main/java/mcbot/Ui.java
#	src/main/java/mcbot/task/Task.java

Commit:
Add more JavaDocs for classes that are not documented previously.

Commit:
Merge branch 'add-gradle-support' into branch-A-Gradle

* add-gradle-support:
  Add Gradle support

Commit:
Add gradle support

Commit:
Merge branch 'branch-A-Gradle'

* branch-A-Gradle:
  Add gradle support
  Add Gradle support

Commit:
Add support checkstyle and edit code that violate the code style (except trailing whitespaces).

Commit:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Add support checkstyle and edit code that violate the code style (except trailing whitespaces).

paing-oo commented on src/main/java/mcbot/McBot.java:
```diff
@@ -0,0 +1,235 @@
+package mcbot;
```
I like how the import statements are listed explicitly and how the ordering is consistent throughout the different classes.

paing-oo commented on src/main/java/mcbot/McBot.java:
```diff
@@ -0,0 +1,235 @@
+                parser.readFullCommand();
+                String keyCommand = parser.getKeyCommand();
+                ui.printFrame();
+                switch(keyCommand) {
```
I like how you make the correct indentations for the switch cases.

paing-oo commented on src/main/java/mcbot/Parser.java:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * To parse the entire line of input. 
+     */
+    public void readFullCommand() {
```
I like how the methods are named in a 'verb' format.

paing-oo commented on src/main/java/mcbot/McBot.java:
```diff
@@ -0,0 +1,235 @@
+        boolean isRunning = true;
+        ui.welcomeLine();
+        Parser parser = new Parser();
+        while (isRunning) {
```
I like how the boolean variable 'isRunning' is named to sound boolean.

Commit:
Fix white-space after bracket.

Commit:
Add GUI support for McBot chat

Commit:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI support for McBot chat

CS2103Lerulatus commented on src/main/java/mcbot/McBot.java:
```diff
@@ -0,0 +1,236 @@
+                        tasks.add(t);
+                        storage.appendData(t);
+                        ui.addEventLine(t, tasks.size());
+                    } catch (InvalidCommandException e) {
+                        ui.printError(e);
+                    } catch (ArrayIndexOutOfBoundsException e) {
+                        ui.taskError("eventFormat");
+                    } catch (DateTimeParseException e) {
+                        ui.taskError("datetimeFormat");
+                    }
+                    break;
+                case "delete":
+                    try {
+                        String numStr = parser.getDetails();
+                        int num = Integer.parseInt(numStr);
+                        if (num < 1 || num > tasks.size()) {
+                            throw new McBotException("Boi, I can't delete a number that ain't on the list");
+                        }
+                        Task t = tasks.get(num - 1);
+                        tasks.remove(num - 1);
+                        storage.updateData(tasks.getList());
+                        ui.deleteLine(t, tasks.size());
+                    } catch (InvalidCommandException | ArrayIndexOutOfBoundsException e) {
+                        ui.deleteError("empty");
+                    } catch (McBotException e) {
+                        ui.printError(e);
+                    } catch (NumberFormatException e) {
+                        ui.deleteError("notInteger");
+                    }
+                    break;
+                case "find":
+                    try {
+                        String taskName = parser.getDetails();
+                        tasks.find(taskName, ui);
+                    } catch (InvalidCommandException | ArrayIndexOutOfBoundsException e) {
+                        ui.taskError("emptyFindTask");
+                    }
+                    break;
+                default:
+                    throw new InvalidCommandException("I don't understand a word ye're sayin'");
```
——— For PR review ———
Might I suggest more usage of OOP-styles -- create a command object for each command that stores its actions and the aliases that call forth it.

CS2103Lerulatus commented on src/main/java/mcbot/Storage.java:
```diff
@@ -0,0 +1,139 @@
+package mcbot;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.ArrayList;
+import java.util.Scanner;
```
——— For PR review ———
Nice and clear individual imports

Commit:
Add assert

Using assert feature to to document
important assumptions that should hold at various points in the code.

Commit:
Refactor code

McBot and Ui Class no longer in use. Numerous arrowhead code were found
and some methods were too long.

McBot and Ui classes were deleted. Arrowhead codes were rectified.
Methods that were too long were cut. Overall, code quality was
improved drastically.

AD-NAP committed:
Merge pull request #1 from AD-NAP/branch-A-CodeQuality

Refactor code

Commit:
Merge branch 'master' of https://github.com/AD-NAP/ip

* 'master' of https://github.com/AD-NAP/ip:
  Refactor code

Commit:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assert

# Conflicts:
#	src/main/java/mcbot/McBotGui.java

Commit:
Add continuous integration

Commit:
Add detecting anomaly feature

Users could add multiple event with the same date and time.

Each time user adds an event, it will ensure it does not clash with
another event that has the same date and time.

Commit:
Update GUI

GUI lack colour and did not go well with user image.

Different dialog box colour depending on the user or chatbot is
responding. Better user friendliness to be able to
differentiate between the two users.

AD-NAP committed:
Set theme jekyll-theme-cayman

Commit:
Add User Guide

User guide for new users to understand how to use the chat bot.

Commit:
Modified User Guide

Commit:
Modified User Guide

Commit:
Modified User Guide

Commit:
Update User Guide

Minor code quality adjustment

Commit:
Minor code quality adjustments

Commit:
Update GUI

Placed a background colour with text bubbles. Images are also
centered.

Added missing JavaDocs.

Commit:
Code quality changes

Minor changes to code to improve code quality and pass
checkstyles.

Commit:
Update CI

CI checking inputs vs outputs text UI but is no longer in use.

Let's update CI file to only check using gradle.
